src/peergos/server/tests/simulation/AccessControl.java

```diff
@@ -0,0 +1,90 @@
+package peergos.server.tests.simulation;
+
+import java.nio.file.Path;
+import java.util.*;
+
+public interface AccessControl {
+
+    String getOwner(Path path);
+
+    List<String> get(Path path, FileSystem.Permission permission);
+
+    void add(Path path, String reader, FileSystem.Permission permission);
+
+    void remove(Path path, String reader, FileSystem.Permission permission);
+
+
+    /**
+     * Model owners, reader and writers
+     */
+    class AccessControlUnit {
+        private Map<Path, List<String>> allowed = new HashMap<>();
+
+        List<String> getAllowed(Path path) {
+            return allowed.computeIfAbsent(path, p -> new ArrayList<>());
+        }
+
+        void addAllowed(Path path, String user) {
+            getAllowed(path).add(user);
+        }
+
+        void removeAllowed(Path path, String user) {
+            getAllowed(path).remove(user);
+        }
+    }
+
+    default boolean can(Path path, String user, FileSystem.Permission permission) {
+        return getOwner(path).equals(user) || get(path, permission).contains(user);
+    }
+
+    class MemoryImpl implements AccessControl {
+        private Map<Path, String> owners = new HashMap<>();
+        public AccessControlUnit readers = new AccessControlUnit();
```

Comment:
Another thing we'll eventually is transitive read access (which we definitely want) and revocation. If A owns file X and A grants B R(X), and then b grants C R(X). Then A revokes R(X) from B, that also revokes it from anyone downstream from B, including C.